### PR TITLE
[dev] PBIs AB#61450 AB#61451 AB#61455: ForwardedRefs, Button Type Prop, and Select Bug Fixes

### PR DESCRIPTION
**Pull request by cbrewer1689**

### CODE REVIEW

File: src/inputs/datePickerInput/datePickerInput.jsx [hunk 1]
```diff
@@ -361,7 +363,7 @@ class DatePickerInput extends React.PureComponent {
 
         if (isNotDisabled) {
             // eslint-disable-next-line no-underscore-dangle, react/no-find-dom-node
-            ReactDOM.findDOMNode(this._datePickerInput._input).focus();
+            this.datePickerInput.current.inputElement.focus();
```
Comment (jeoffm) [inline src/inputs/datePickerInput/datePickerInput.jsx:366]: looks like the eslint directive above is no longer necessary

File: src/inputs/input/input.jsx [hunk 2]
```diff
@@ -133,15 +132,25 @@ class Input extends React.PureComponent {
...
+            loading,
+            mask,
+        } = this.props;
+
         const type = this.getType();
 
         if (isString(icon) || isObject(icon) || loading || type === 'number') {
             // eslint-disable-next-line react/no-find-dom-node, no-underscore-dangle
-            const inputTop = ReactDOM.findDOMNode(this._input).offsetTop;
+            const inputTop = mask ?
```
Comment (jeoffm) [inline src/inputs/input/input.jsx:151]: eslint directive not needed

File: src/inputs/input/input.jsx [hunk 3]
```diff
@@ -150,7 +159,11 @@ class Input extends React.PureComponent {
 
         if (autoFocus) {
             // eslint-disable-next-line react/no-find-dom-node, no-underscore-dangle
-            ReactDOM.findDOMNode(this._input).focus();
+            if (mask) {
```
Comment (jeoffm) [inline src/inputs/input/input.jsx:162]: eslint directive

File: src/inputs/input/input.jsx [hunk 4]
```diff
@@ -362,8 +381,11 @@ class Input extends React.PureComponent {
         if (isFunction(onChange)) {
             onChange(value);
         } else {
-            // eslint-disable-next-line no-underscore-dangle
-            this._input.value = value;
+            if (mask) {
```
Comment (jeoffm) [inline src/inputs/input/input.jsx:384]: looks like this should be moved up to the else line (ie, it s/b an else-if)

File: src/inputs/textArea/textArea.jsx [hunk 5]
```diff
@@ -108,9 +109,9 @@ class TextArea extends React.Component {
             value,
         } = this.props;
 
-        if (autoHeight && this.textArea) {
+        if (autoHeight && this.textArea.current) {
             // eslint-disable-next-line react/no-find-dom-node
-            const textArea = ReactDOM.findDOMNode(this.textArea);
+            const textArea = this.textArea.current;
```
Comment (jeoffm) [inline src/inputs/textArea/textArea.jsx:114]: eslint directive

File: src/inputs/textArea/textArea.jsx [hunk 6]
```diff
@@ -124,9 +125,9 @@ class TextArea extends React.Component {
             }, 150);
         }
 
-        if (autoFocus && this.textArea) {
+        if (autoFocus && this.textArea.current) {
             // eslint-disable-next-line react/no-find-dom-node
-            ReactDOM.findDOMNode(this.textArea).focus();
+            this.textArea.current.focus();
```
Comment (jeoffm) [inline src/inputs/textArea/textArea.jsx:130]: eslint

File: src/inputs/textArea/textArea.jsx [hunk 7]
```diff
@@ -156,9 +158,9 @@ class TextArea extends React.Component {
 
         if (autoHeight && isFunction(onAutoHeightResized)) {
             // eslint-disable-next-line react/no-find-dom-node
-            const textArea = ReactDOM.findDOMNode(this.textArea);
+            const textAreaElement = this.textArea.current;
```
Comment (jeoffm) [inline src/inputs/textArea/textArea.jsx:161]: eslint

**Review by jeoffm — APPROVED** [review]

File: src/inputs/input/input.jsx (hunk 4 above)
Comment (groberts314) [reply src/inputs/input/input.jsx:384]: Maybe, but sometimes keeping the nesting helps readability.

i.e. there are two completely independent things going on here:
1) Is it a "controlled component" or not? (outer `if`/`else`)
2) Are we using an input mask or not (inner `if`/`else` within the outer `else` [not a controlled component])

I guess we could do this to refactor/simplify as you suggest and assuage my readability concerns (and I suspect @cbrewer1689's concerns as well):
```JavaScript
if (isFunction(onChange)) { // if it's a controlled component
    onChange(value);
} else if (mask) { // not a controlled component; using an input mask
    this.input.current.inputElement.value = value;
} else { // not a controlled component; not using an input mask
    this.input.current.value = value;
}
```

File: src/inputs/select/select.jsx [hunk 8]
```diff
@@ -851,9 +851,6 @@ const Select = React.forwardRef(function Select(props, ref) {
                 <div
                     aria-selected={isSelected}
                     className={optionClass}
-                    isDisabled={o.disabled}
-                    isFocused={isFocused}
-                    isSelected={isSelected}
```
Comment (groberts314) [inline src/inputs/select/select.jsx:856]: Thank you!!!!! 👍 🎉 
These console warnings have been driving me _nuts!_ =)

**Review by groberts314 — COMMENTED** [review]
Looks good overall!

Just submitting comments for now (don't wanna approve, but don't wanna formally request 
changes).  I had one question where something maybe doesn't look quite right, but it might well be fine.  If there is feedback on that, I will happily approve.

I also concur with @jeoffm that any ESLINT suppression comments that are no longer needed should ideally be removed.

**Comment by groberts314** [comment]
Just a heads up: At least  chance of conflicts with #353

File: src/inputs/select/select.jsx (hunk 8 above)
Comment (cbrewer1689) [reply src/inputs/select/select.jsx:856]: Me too.

**Comment by cbrewer1689** [comment]
I'll go through your guy's catches and suggestions later today.

File: src/inputs/input/input.jsx (hunk 4 above)
Comment (cbrewer1689) [reply src/inputs/input/input.jsx:384]: I like the suggestion here.

File: src/inputs/input/input.jsx [hunk 9]
```diff
@@ -276,14 +289,20 @@ class Input extends React.PureComponent {
...
         const isDisabled = disable || disabled;
-        // eslint-disable-next-line no-underscore-dangle
-        const { value } = this._input;
+        let value;
+
+        if (mask) {
+            value = this.input.current.inputElement;
+        } else {
+            value = this.input.current;
+        }
```
[the remark this answers is not in the document]
Comment (cbrewer1689) [reply src/inputs/input/input.jsx:305]: I think this is a good catch. I am going to add `.value` and test the number toggles.

File: src/inputs/datePickerInput/datePickerInput.jsx (hunk 1 above)
Comment (cbrewer1689) [reply src/inputs/datePickerInput/datePickerInput.jsx:365]: Fixed.

File: src/inputs/input/input.jsx (hunk 2 above)
Comment (cbrewer1689) [reply src/inputs/input/input.jsx:150]: Fixed.

File: src/inputs/input/input.jsx (hunk 3 above)
Comment (cbrewer1689) [reply src/inputs/input/input.jsx:160]: Fixed.

File: src/inputs/input/input.jsx (hunk 4 above)
Comment (cbrewer1689) [reply src/inputs/input/input.jsx]: Fixed.

File: src/inputs/textArea/textArea.jsx (hunk 5 above)
Comment (cbrewer1689) [reply src/inputs/textArea/textArea.jsx:113]: Fixed.

File: src/inputs/textArea/textArea.jsx (hunk 6 above)
Comment (cbrewer1689) [reply src/inputs/textArea/textArea.jsx:128]: Fixed.

File: src/inputs/input/input.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (cbrewer1689) [reply src/inputs/input/input.jsx]: Fixed.

**Pull request merged**